### DISCUSSION TITLE
Who is using server.url in production?

### DISCUSSION THREAD

**Discussion by jgoux** [ionic-team/capacitor#4080] [Q&A]
Hello all.

We want to use capacitor at my current company to wrap a Next.js application for mobile/electron distribution.

I know server.url isn't meant for production but I also know that some people are using it anyway to load a whole hosted PWA.

We have several requirements that seems to be hard to fulfill using the regular approach :
- We use [WOPI](https://wopi.readthedocs.io/en/latest/) (Office Online in-app integration) which requires to whitelist our app's domain on Microsoft's side to work. But capacitor's domains aren't the same as the hosted web app. I know we could use the `hostname` configuration but it seems to introduce issues on Android's side as it breaks the secure context.
- It seems there is  a lot of issues with [cookies](https://github.com/ionic-team/capacitor/issues/1373#) in Capacitor, and we plan to rely on them for our authentication layer (auth0).

So...on one hand we could just take our working PWA and use server.url and keep going with our lifes, or try to do it the regular way with a lot of frictions, if the thing is even possible.

This is why I'm curious to know if anyone in the community is using `server.url` in production, and if there were any issue being accepted on the stores doing it, particularly on iOS App Store. 😅

Thanks for your time! 👍

**Comment by jayenashar** [comment]
Yes, I am using `server.url` in production, and I didn't have issues with the stores:

https://play.google.com/store/apps/details?id=waverley.transport.maps.transit.timetable
https://apps.apple.com/au/app/id1525088783

**Reply by humbletrad** [reply]
Is that any caveat when working on server.url in production? Good to know earlier.
1. HTTPS is a must (from Android 9)

**Reply by LefanTan** [reply]
How are you setting this up in capacitor.config? I have my server URL set to a production website but only a blank page shows up

**Reply by jayenashar** [reply]
you'll need to connect chrome/safari and see what the issue is.  hopefully this happens in dev so you can connect.

**Reply by konasunny** [reply]
> Yes, I am using `server.url` in production, and I didn't have issues with the stores:
> 
> https://play.google.com/store/apps/details?id=waverley.transport.maps.transit.timetable https://apps.apple.com/au/app/id1525088783

i was trying to use this approach by settig the server.url. as a result i am able to lod all my fiels by the capacitor plugins are not working. How can resolve this issue.

**Reply by jayenashar** [reply]
@konasunny i didn't have any issues using `window.Capacitor.Plugins` in my app

**Comment by bigboypantson** [comment]
I am also interested in using server.url for production, do you have any further updates on your experimenting @jgoux?

**Comment by scottwilson312** [comment]
So far when I try to do this, `capacitor.getPlatform` thinks it is in web mode.... not android. And therefore none of the native plugins work. Does anyone have any answers to this? This seems like a bizarre limitation.

**Reply by jayenashar** [reply]
what version of capacitor?  Geolocation, Storage, and LocalNotifications work for me on 2.4.2

**Reply by scottwilson312** [reply]
2.4.7. I believe plugins would work because they have a web implementation. My custom plugin only works on native Android. When you run Capacitor.getPlatform() on your app does it show as web or native?

**Reply by jayenashar** [reply]
"android"

![image](https://user-images.githubusercontent.com/1137152/120285788-f30c0680-c300-11eb-8662-119c9d0187cd.png)

**Reply by jayenashar** [reply]
LocalNotificaitons does not have a web implementation, btw.

**Reply by mlynch** [reply]
@jayenashar it does have a web implementation https://github.com/ionic-team/capacitor-plugins/blob/main/local-notifications/src/web.ts

**Reply by jayenashar** [reply]
thanks.  maybe i'm on an old version (2.4.2) that doesn't have it?  i get an error when i try to use it on web.

**Comment by KoenLav** [comment]
We are using server.url (given a similar scenario) with 2.4.* and haven't experienced any issues so far.

Any specific reasons from the Ionic team to not consider server.url to be a first class citizen?

**Reply by jcesarmobile** [reply]
If you check [point 4.7 ](https://developer.apple.com/app-store/review/guidelines/#third-party-software)of Apple Store Review Guidelines
> Apps may contain or run code that is not embedded in the binary
So that's ok, you can point your app to an url without problems

But then:
> 4.7.1 Software offered under this rule must:
> only use capabilities available in a standard WebKit view (e.g. it must open and run natively in Safari without modifications or additional software); and use WebKit and JavaScript Core to run third-party software and should not attempt to extend or expose native platform APIs to third-party software;

That's a grey area, Capacitor plugins don't work in a standard WebKit view, well, most of them will fallback to the web version, so you can argue that it "still works" if Apple rejects your app because of that. Might be good if you don't expose the plugins to third-party software.

I think Google had similar rules, but not sure if they removed then now that they allow to submit PWAs.

Anyway, the only reason we added the server.url config option was because of live reload servers, which is used for development purposes, so we don't recommend to use it in production apps because as I said, it can lead to app rejections, so use at your own risk.

**Reply by echarbeto** [reply]
Hello,
The server.url with a domain will allow to mount an iframe that, for security reasons, does not accept "locahost" as a valid domain within the x-frame-options.

Is there any other alternative?

**Reply by michael-scheurer** [reply]
Suggestion: Just point out this possibility in the docs with a warning note that the app might be rejected by Apple. In other projects I also just load an url over WebViews and wrote some easy API. My apps have never been rejected from Apple.

Loading the app from url brings a lot of advantages. Especially it faciliates the deplyment process.Because you don't need to update the native apps in most cases.

**Reply by jayenashar** [reply]
but then who will use ionic appflow?

**Comment by berkayyildiz** [comment]
@KoenLav @jayenashar We are also planning to use server.url for production but it seems like we can't get updates like web. We are using angular and currently our implementation for web is basically like this but when we use server.url capacitor does not get any updates. How do you get updates on your app?

```
if (this.swUpdate.isEnabled)
      this.swUpdate.available.subscribe(() => {
            // get confirm and update
            this.swUpdate.activateUpdate().then(() => document.location.reload())
      })
```

**Reply by jayenashar** [reply]
i'm not using service workers.  your issue looks like it's specific to how you use service workers.

**Reply by KoenLav** [reply]
We ran into similar issues with service workers.

**Reply by berkayyildiz** [reply]
> We ran into similar issues with service workers.

How did you solve them. How both of you getting updates in the app? When we use server.url it fetches for one time, after even when production updated, app do not fetch the new version.

**Reply by KoenLav** [reply]
We didn't, ended up instrumenting the backend to send an app version which would cause the frontend to (allow the user to) reload whenever it didn't match.

**Reply by berkayyildiz** [reply]
But how are you update it, what is the code for it?

**Comment by scottwilson312** [comment]
I'm using it in production in both iOS and Android app stores. It works great. You need to go into node_modules and edit out the little toast that says "using server url...."

**Reply by jayenashar** [reply]
that toast only appears in dev for me, not prod

**Comment by goforu** [comment]
Hello I'm using server url for production, in the server side, we are using PWA. But app can never get updated. Anyone have a solution for this?

**Comment by snimavat** [comment]
Interested in knowing the experience of other people who are using this in production, and if it works flowlessly and can update the assets on launch etc.

**Reply by jayenashar** [reply]
not sure what you mean by "update the assets on launch".  if the assets are on the server.url then updating them is unrelated to launching the app

**Comment by scottwilson312** [comment]
I'm using this just fine in production, both iOS and Android. No issues with the app store. I can deploy code to the pointed at server and update my app that way. You do however need to go into (for android for example) node_modules and fork capacitor so that it doesnt pop up the toast saying "using server..."

**Reply by jayenashar** [reply]
that toast is only in debug mode.  when you build for production it doesn't appear.

**Reply by scottwilson312** [reply]
How do you build for production? If I produce a signed production build for the app store, it still appears unless I comment it out. I should note I'm using an older capacitor version though

**Reply by jayenashar** [reply]
for android i do:

```sh
npx ionic capacitor sync --prod --platform=android
cd android
./gradlew --no-daemon assembleRelease
```

for ios someone else takes care of it but they are doing it within xcode.

**Reply by jayenashar** [reply]
i see the file you mention specifically has `if (BuildConfig.DEBUG)` so it should not have the toast in prod.

**Reply by jayenashar** [reply]
or `if bridge!.isDevMode() &&` on ios

**Reply by scottwilson312** [reply]
thank you!!!

**Comment by 27pchrisl** [comment]
How are people dealing with the user potentially opening your app offline, and then the app never being able to recover. Or the user going offline during the experience, and sort of crashing the app because it can't load from the server?

**Reply by jayenashar** [reply]
To clarify, you mean that once the app is opened offline, the user needs to force-quit to open the app online later?

**Reply by 27pchrisl** [reply]
Yes, exactly that.

**Reply by 27pchrisl** [reply]
As the user experience would be to just get a white screen, which I'm surprised Apple don't reject your app for.

**Reply by jayenashar** [reply]
once you're in, you're in.  the beauty of requiring an internet connection 😁 

actually our app doesn't close the splash screen with a timeout.  we have some code to stop the splash screen once the original page is fully loaded.  so the user doesn't see a white screen.  just stuck on the splash screen.

**Comment by riderx** [comment]
hey guys, for me server-url is not mean for production.

That because of Apple and Google rules.
For Apple:
https://developer.apple.com/app-store/review/guidelines/
in "4.7 HTML5 Games, Bots, etc."
Apps may contain or run code that is not embedded in the binary[...]
only use capabilities available in a standard WebKit view.

but also
4.2 Minimum Functionality
Your app should include features, content, and UI that elevate it beyond a repackaged website. 

So for me it work until Apple find out.
who don't seems a good business idea 😅

For Google rules are less clear and they seems usually more open.
https://support.google.com/googleplay/android-developer/answer/9888379?hl=en#zippy=%2Cexamples-of-common-violations

But i wouldn't bet on this for long therm.

I believe that why server-url is not first-class citizen.
The golden way for ionic is AppFlow, that a good service they provide, and do the job well.

There alternative now, same as you would do in electron you can use plugin to fech your updates and have your code working offline.
check this:
https://github.com/Cap-go/capacitor-updater

**Reply by jayenashar** [reply]
4.7 looks fine to me...  it says "may" not "may not"

4.2 isn't that much of a concern.  i've had that flagged upon review but pointing out other popular apps doing the same was enough to get by (e.g. airbnb, uber, starbucks, pinterest, etc.)

Ionic AppFlow doesn't really do anything to change your app to adhere to 4.2 or 4.7...

**Reply by riderx** [reply]
For 4.7, the point i see problematic is `only use capabilities available in a standard WebKit view.` so for them if you app content is external then no more than just web api is allowed.
Who can be good for some, but not for all of us.
For 4.2 that always disputable too, what I see is more you expose yourself to risk, more chance to get kick out when the wind direction change in Apple.

**Reply by jayenashar** [reply]
4.7 what more do you need for code not embedded in the binary?

4.2 if apple wants to kick you out, they will find a reason.  if you are no threat, they won't spend the time/effort.  https://www.timetoplayfair.com/ https://techcrunch.com/2022/09/01/apple-settles-lawsuit-with-developer-over-app-store-rejections-and-scams/

**Reply by riderx** [reply]
4.7, so your JS from this URL shouldn't call capacitor plugin.

4.2 yes that why I choose to have less reason as possible.

**Reply by jayenashar** [reply]
4.7 why not?  capacitor plugin is code embedded in the binary

4.2 even if you have no reason (like the links i gave) they will make a reason.

**Reply by riderx** [reply]
4.7 :  for the exact reason i said first time...
Apps may contain or run code that is not embedded in the binary[...]
only use capabilities available in a standard WebKit view.

4.2 it's just a statistic approach, more you play the way they expect more change to survive.

**Reply by jayenashar** [reply]
4.7 i still don't get it.  the app runs both code embedded and not embedded in the binary.  the code not embedded in the binary only uses capabilities available in a standard WebKit view...  or are you saying there is something non-standard about the capacitor webkit view?

4.2 not really.  if i flip a coin 100 times and 100 times it lands on the same side, what is the chance that the 101th flip will land on the same side?  you may think it's 50% because past flips don't affect future flips but it's not.  the coin (or the flip) is obviously rigged.  it's not statistics, it's a rigged game.  you either make money for apple or they find a reason.  it has nothing to do with following the rules.

**Reply by riderx** [reply]
there no point to continue this, i hope that will be fine for you like that.

**Reply by jayenashar** [reply]
that's fine for me.  what was your goal to start this?  to get more business for ionic appflow?

**Reply by riderx** [reply]
I'm just helping the community, I saw the same question on the discord and answered. 
What I believe, is the more safe way. I deal with Apple review since 10 years now.

Furthermore, I have made [capacitor-updater](https://github.com/Cap-go/capacitor-updater), who allow you to don't need AppFlow. 
So, I'm not helping them to get more business.

**Reply by jayenashar** [reply]
i think better you point them to @jcesarmobile 's answer https://github.com/ionic-team/capacitor/discussions/4080#discussioncomment-873991

**Reply by riderx** [reply]
Well, that exactly what I was pointing out in my first message.

**Comment by hhimanshu** [comment]
I tried to use `server.url` but failed in my attempt. I started a new discussion at https://github.com/ionic-team/capacitor/discussions/6546

Hoping to learn more from people here and fix this issue
